File: frontend/src/lib/production.ts
import { apiClient } from "@/lib/api";

export type PlanStatus = "DRAFT" | "CONFIRMED" | "IN_PROGRESS" | "COMPLETED" | "CANCELLED";
export type OrderStatus = "PLANNED" | "RELEASED" | "IN_PROGRESS" | "COMPLETED" | "CANCELLED";
export type DowntimeReason =
  | "BREAKDOWN"
  | "MAINTENANCE"
  | "CHANGEOVER"
  | "MATERIAL_SHORTAGE"
  | "QUALITY_HOLD"
  | "OTHER";

export interface ProductionPlan {
  id: string;
  plan_no: string;
  name: string;
  description?: string;
  status: PlanStatus;
  planned_date: string;
  created_by?: string;
}

export interface PlanLine {
  id: string;
  plan_id: string;
  line_no: number;
  product_id: string;
  product_name?: string;
  product_sku?: string;
  recipe_id: string;
  recipe_name?: string;
  recipe_version?: string;
  target_quantity: number;
  uom: string;
  target_warehouse_id: string;
  target_warehouse_name?: string;
  planned_date?: string;
  notes?: string;
}

export interface ProductionPlanDetail extends ProductionPlan {
  lines: PlanLine[];
}

export interface ProductionOrder {
  id: string;
  order_no: string;
  plan_line_id?: string;
  product_id: string;
  product_name?: string;
  product_sku?: string;
  recipe_id: string;
  recipe_name?: string;
  planned_quantity: number;
  actual_quantity?: number;
  batch_no?: string;
  uom: string;
  status: OrderStatus;
  target_warehouse_id: string;
  target_warehouse_name?: string;
  source_warehouse_id?: string;
  scheduled_start: string;
  scheduled_end: string;
  actual_start?: string;
  actual_end?: string;
  created_by?: string;
  notes?: string;
}

export interface MaterialConsumption {
  id: string;
  material_id: string;
  material_name?: string;
  material_code?: string;
  lot_id?: string;
  lot_number?: string;
  standard_quantity: number;
  actual_quantity: number;
  variance_quantity: number;
  unit: string;
  source_warehouse_id: string;
  source_warehouse_name?: string;
}

export interface FGReceipt {
  id: string;
  product_id: string;
  quantity: number;
  uom: string;
  lot_number?: string;
  batch_no?: string;
  warehouse_id: string;
  warehouse_name?: string;
}

export interface DowntimeLog {
  id: string;
  production_order_id: string;
  machine_id?: string;
  reason: DowntimeReason;
  start_time: string;
  end_time?: string;
  duration_minutes?: number;
  notes?: string;
}

export interface OEEMetrics {
  order_id: string;
  order_no: string;
  scheduled_minutes: number;
  actual_minutes?: number;
  downtime_minutes: number;
  availability: number;
  performance: number;
  quality: number;
  oee: number;
}

export interface ProductionOrderDetail extends ProductionOrder {
  consumptions: MaterialConsumption[];
  fg_receipts: FGReceipt[];
  downtime_logs: DowntimeLog[];
  oee?: OEEMetrics;
}

export interface ProductionReportRow {
  date: string;
  order_no: string;
  product_name: string;
  product_sku: string;
  planned_quantity: number;
  actual_quantity?: number;
  variance?: number;
  variance_pct?: number;
  status: OrderStatus;
  batch_no?: string;
  downtime_minutes: number;
}

export interface DowntimeSummaryRow {
  reason: string;
  count: number;
  total_minutes: number;
  machine_id?: string;
}

export interface ProductionSummary {
  total_orders: number;
  completed_orders: number;
  in_progress_orders: number;
  planned_quantity_total: number;
  actual_quantity_total: number;
  overall_variance: number;
  total_downtime_minutes: number;
  avg_oee?: number;
  orders: ProductionReportRow[];
  downtime_summary: DowntimeSummaryRow[];
}

export interface PlanCreate {
  plan_no: string;
  name: string;
  description?: string;
  planned_date: string;
}

export interface PlanLineCreate {
  line_no: number;
  product_id: string;
  recipe_id: string;
  target_quantity: number;
  uom: string;
  target_warehouse_id: string;
  planned_date?: string;
  notes?: string;
}

export interface OrderCreate {
  order_no: string;
  plan_line_id?: string;
  product_id: string;
  recipe_id: string;
  planned_quantity: number;
  uom: string;
  batch_no?: string;
  target_warehouse_id: string;
  source_warehouse_id?: string;
  scheduled_start: string;
  scheduled_end: string;
  notes?: string;
}

export interface CompleteOrderRequest {
  actual_quantity: number;
  fg_lot_number?: string;
  fg_expiry_date?: string;
  source_warehouse_id?: string;
  consume_materials?: boolean;
  notes?: string;
}

export const productionApi = {
  // Plans
  async listPlans(params?: { status?: PlanStatus; skip?: number; limit?: number }): Promise<ProductionPlan[]> {
    const res = await apiClient.get<ProductionPlan[]>("/api/v1/production/plans/", { params }).then((r) => r.data);
    return res.data;
  },
  async getPlan(id: string): Promise<ProductionPlanDetail> {
    const res = await apiClient.get<ProductionPlanDetail>(`/api/v1/production/plans/${id}`).then((r) => r.data);
    return res.data;
  },
  async createPlan(data: PlanCreate): Promise<ProductionPlan> {
    const res = await apiClient.post<ProductionPlan>("/api/v1/production/plans/", data).then((r) => r.data);
    return res.data;
  },
  async confirmPlan(id: string): Promise<ProductionPlan> {
    const res = await apiClient.post<ProductionPlan>(`/api/v1/production/plans/${id}/confirm`).then((r) => r.data);
    return res.data;
  },
  async addPlanLine(planId: string, data: PlanLineCreate): Promise<PlanLine> {
    const res = await apiClient.post<PlanLine>(`/api/v1/production/plans/${planId}/lines`, data).then((r) => r.data);
    return res.data;
  },
  async deletePlanLine(planId: string, lineId: string): Promise<void> {
    await apiClient.delete(`/api/v1/production/plans/${planId}/lines/${lineId}`).then((r) => r.data);
  },

  // Orders
  async listOrders(params?: { status?: OrderStatus; product_id?: string; skip?: number; limit?: number }): Promise<ProductionOrder[]> {
    const res = await apiClient.get<ProductionOrder[]>("/api/v1/production/orders/", { params }).then((r) => r.data);
    return res.data;
  },
  async getOrder(id: string): Promise<ProductionOrderDetail> {
    const res = await apiClient.get<ProductionOrderDetail>(`/api/v1/production/orders/${id}`).then((r) => r.data);
    return res.data;
  },
  async createOrder(data: OrderCreate): Promise<ProductionOrder> {
    const res = await apiClient.post<ProductionOrder>("/api/v1/production/orders/", data).then((r) => r.data);
    return res.data;
  },

  // Execution
  async releaseOrder(id: string): Promise<ProductionOrder> {
    const res = await apiClient.post<ProductionOrder>(`/api/v1/production/orders/${id}/release`).then((r) => r.data);
    return res.data;
  },
  async startOrder(id: string): Promise<ProductionOrder> {
    const res = await apiClient.post<ProductionOrder>(`/api/v1/production/orders/${id}/start`).then((r) => r.data);
    return res.data;
  },
  async pauseOrder(id: string, data: { machine_id?: string; reason: DowntimeReason; notes?: string }): Promise<DowntimeLog> {
    const res = await apiClient.post<DowntimeLog>(`/api/v1/production/orders/${id}/pause`, data).then((r) => r.data);
    return res.data;
  },
  async resumeOrder(id: string, notes?: string): Promise<DowntimeLog> {
    const res = await apiClient.post<DowntimeLog>(`/api/v1/production/orders/${id}/resume`, { notes }).then((r) => r.data);
    return res.data;
  },
  async completeOrder(id: string, data: CompleteOrderRequest): Promise<ProductionOrderDetail> {
    const res = await apiClient.post<ProductionOrderDetail>(`/api/v1/production/orders/${id}/complete`, data).then((r) => r.data);
    return res.data;
  },
  async cancelOrder(id: string): Promise<ProductionOrder> {
    const res = await apiClient.post<ProductionOrder>(`/api/v1/production/orders/${id}/cancel`).then((r) => r.data);
    return res.data;
  },
  async getOEE(id: string): Promise<OEEMetrics> {
    const res = await apiClient.get<OEEMetrics>(`/api/v1/production/orders/${id}/oee`).then((r) => r.data);
    return res.data;
  },

  // Reports
  async getReport(params: { date_from: string; date_to: string; product_id?: string }): Promise<ProductionSummary> {
    const res = await apiClient.get<ProductionSummary>("/api/v1/production/reports/summary", { params }).then((r) => r.data);
    return res.data;
  },
};

File: backend/app/schemas/production.py
from pydantic import BaseModel
from typing import Optional, List
from decimal import Decimal
from datetime import datetime
import uuid

from app.models.production import (
    ProductionPlanStatus, ProductionOrderStatus, DowntimeReason,
)
from app.models.finance import OperationalPostingStatus


# ── Production Plan Line ──────────────────────────────────────────────────────

class PlanLineCreate(BaseModel):
    line_no: int
    product_id: uuid.UUID
    recipe_id: uuid.UUID
    target_quantity: Decimal
    uom: str = "KG"
    target_warehouse_id: uuid.UUID
    planned_date: Optional[datetime] = None
    notes: Optional[str] = None


class PlanLineRead(PlanLineCreate):
    id: uuid.UUID
    plan_id: uuid.UUID
    product_name: Optional[str] = None
    product_sku: Optional[str] = None
    recipe_name: Optional[str] = None
    recipe_version: Optional[str] = None
    target_warehouse_name: Optional[str] = None
    model_config = {"from_attributes": True}


# ── Production Plan ───────────────────────────────────────────────────────────

class ProductionPlanCreate(BaseModel):
    plan_no: str
    name: str
    description: Optional[str] = None
    planned_date: datetime


class ProductionPlanUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    planned_date: Optional[datetime] = None


class ProductionPlanRead(BaseModel):
    id: uuid.UUID
    plan_no: str
    name: str
    description: Optional[str] = None
    status: ProductionPlanStatus
    planned_date: datetime
    created_by: Optional[uuid.UUID] = None
    model_config = {"from_attributes": True}


class ProductionPlanDetailRead(ProductionPlanRead):
    lines: List[PlanLineRead] = []


# ── Production Order ──────────────────────────────────────────────────────────

class ProductionOrderCreate(BaseModel):
    order_no: str
    plan_line_id: Optional[uuid.UUID] = None
    product_id: uuid.UUID
    recipe_id: uuid.UUID
    planned_quantity: Decimal
    uom: str = "KG"
    batch_no: Optional[str] = None
    target_warehouse_id: uuid.UUID
    source_warehouse_id: Optional[uuid.UUID] = None
    scheduled_start: datetime
    scheduled_end: datetime
    notes: Optional[str] = None


class ProductionOrderRead(BaseModel):
    id: uuid.UUID
    order_no: str
    plan_line_id: Optional[uuid.UUID] = None
    product_id: uuid.UUID
    recipe_id: uuid.UUID
    planned_quantity: Decimal
    actual_quantity: Optional[Decimal] = None
    batch_no: Optional[str] = None
    uom: str
    status: ProductionOrderStatus
    target_warehouse_id: uuid.UUID
    source_warehouse_id: Optional[uuid.UUID] = None
    scheduled_start: datetime
    scheduled_end: datetime
    actual_start: Optional[datetime] = None
    actual_end: Optional[datetime] = None
    created_by: Optional[uuid.UUID] = None
    notes: Optional[str] = None
    product_name: Optional[str] = None
    product_sku: Optional[str] = None
    recipe_name: Optional[str] = None
    target_warehouse_name: Optional[str] = None
    model_config = {"from_attributes": True}


# ── Material Consumption ──────────────────────────────────────────────────────

class MaterialConsumptionRead(BaseModel):
    id: uuid.UUID
    production_order_id: uuid.UUID
    material_id: uuid.UUID
    lot_id: Optional[uuid.UUID] = None
    standard_quantity: Decimal
    actual_quantity: Decimal
    variance_quantity: Decimal
    unit: str
    source_warehouse_id: uuid.UUID
    stock_movement_id: Optional[uuid.UUID] = None
    posting_batch_id: Optional[uuid.UUID] = None
    journal_entry_id: Optional[uuid.UUID] = None
    accounting_status: Optional[OperationalPostingStatus] = None
    posting_error: Optional[str] = None
    notes: Optional[str] = None
    material_name: Optional[str] = None
    material_code: Optional[str] = None
    lot_number: Optional[str] = None
    source_warehouse_name: Optional[str] = None
    model_config = {"from_attributes": True}


# ── Finished Goods Receipt ────────────────────────────────────────────────────

class FinishedGoodsReceiptRead(BaseModel):
    id: uuid.UUID
    production_order_id: uuid.UUID
    product_id: uuid.UUID
    quantity: Decimal
    uom: str
    lot_number: Optional[str] = None
    batch_no: Optional[str] = None
    warehouse_id: uuid.UUID
    stock_movement_id: Optional[uuid.UUID] = None
    posting_batch_id: Optional[uuid.UUID] = None
    journal_entry_id: Optional[uuid.UUID] = None
    accounting_status: Optional[OperationalPostingStatus] = None
    posting_error: Optional[str] = None
    notes: Optional[str] = None
    warehouse_name: Optional[str] = None
    model_config = {"from_attributes": True}


# ── Downtime ──────────────────────────────────────────────────────────────────

class DowntimeLogCreate(BaseModel):
    machine_id: Optional[str] = None
    reason: DowntimeReason = DowntimeReason.OTHER
    start_time: Optional[datetime] = None
    notes: Optional[str] = None


class DowntimeLogClose(BaseModel):
    end_time: Optional[datetime] = None
    notes: Optional[str] = None


class DowntimeLogRead(BaseModel):
    id: uuid.UUID
    production_order_id: uuid.UUID
    machine_id: Optional[str] = None
    reason: DowntimeReason
    start_time: datetime
    end_time: Optional[datetime] = None
    duration_minutes: Optional[int] = None
    notes: Optional[str] = None
    model_config = {"from_attributes": True}


# ── OEE ───────────────────────────────────────────────────────────────────────

class OEEMetrics(BaseModel):
    order_id: uuid.UUID
    order_no: str
    scheduled_minutes: float
    actual_minutes: Optional[float] = None
    downtime_minutes: float
    availability: float
    performance: float
    quality: float
    oee: float


# ── Execution requests ────────────────────────────────────────────────────────

class CompleteOrderRequest(BaseModel):
    actual_quantity: Decimal
    fg_lot_number: Optional[str] = None
    fg_expiry_date: Optional[str] = None
    source_warehouse_id: Optional[uuid.UUID] = None
    consume_materials: bool = True
    notes: Optional[str] = None


# ── Production Order Detail (full) ────────────────────────────────────────────

class ProductionOrderDetailRead(ProductionOrderRead):
    consumptions: List[MaterialConsumptionRead] = []
    fg_receipts: List[FinishedGoodsReceiptRead] = []
    downtime_logs: List[DowntimeLogRead] = []
    oee: Optional[OEEMetrics] = None


# ── Reports ───────────────────────────────────────────────────────────────────

class ProductionReportRow(BaseModel):
    date: str
    order_no: str
    product_name: str
    product_sku: str
    planned_quantity: Decimal
    actual_quantity: Optional[Decimal]
    variance: Optional[Decimal]
    variance_pct: Optional[float]
    status: ProductionOrderStatus
    batch_no: Optional[str]
    downtime_minutes: float


class DowntimeSummaryRow(BaseModel):
    reason: str
    count: int
    total_minutes: float
    machine_id: Optional[str]


class ProductionSummary(BaseModel):
    total_orders: int
    completed_orders: int
    in_progress_orders: int
    planned_quantity_total: Decimal
    actual_quantity_total: Decimal
    overall_variance: Decimal
    total_downtime_minutes: float
    avg_oee: Optional[float]
    orders: List[ProductionReportRow]
    downtime_summary: List[DowntimeSummaryRow]
